cmd/acb/commands/exec/exec.go

```diff
@@ -201,55 +200,67 @@ var Command = cli.Command{
+		if err != nil {
+			return err
+		}
+
 		versionInUse := graph.FindVersion(template.GetData())
-		shouldIncludeAlias := versionInUse == "" || versionInUse >= "v1.1.0"
+		shouldIncludeAlias := versionInUse >= "v1.1.0"
 
+		var alias *graph.Alias
 		var task *graph.Task
 		if shouldIncludeAlias {
 			log.Printf("Alias support enabled for version >= 1.1.0, please see https://aka.ms/acr/tasks/task-aliases for more information.")
-			// Generate the base task file without resolving environment variables.
-			task, err = graph.NewTaskFromBytes(template.GetData(), true)
-			if err != nil {
-				return err
-			}
+			// separate alias and remaining data
+			aliasData, remainingData := graph.SeparateAliasFromRest(template.GetData())
+			// render alias data
+			values := templating.GetTaskRenderObject(renderOpts)
+			renderedAlias, err := templating.NewEngine().RenderGoTemplate(uuid.New().String(), string(aliasData), values)
+			if err == nil {
+				aliasData = []byte(renderedAlias)
 
-			// Remarshal functional task file to resolve templating
-			var fromTask []byte
-			fromTask, err = yaml.Marshal(*task)
-			if err != nil {
-				return err
+				// Preprocess the task to replace all aliases based on the alias sources.
+				post, _alias, _, aliasErr := graph.SearchReplaceAlias(template.GetData(), aliasData, remainingData)
+				alias = _alias
+				if aliasErr != nil {
+					log.Printf("unable to search/replace alias in task %s due to error %+v", post, aliasErr)
+				} else {
+					// update the template.Data
+					template.Data = post
+				}
+			} else {
+				log.Printf("unable to render alias %s due to error %+v", string(aliasData), err)
 			}
-			template.Data = fromTask
 		}
 
 		rendered, err := templating.LoadAndRenderSteps(ctx, template, renderOpts)
 		if err != nil {
 			return err
 		}
-
-		if debug {
+		if ctx.Value("debug").(bool) {
 			log.Println("Rendered template:")
 			log.Println(rendered)
 		}
 
-		var credentials []*graph.RegistryCredential
-		// Add all creds provided by the user in the --credential flag
-		for _, credString := range creds {
-			var cred *graph.RegistryCredential
-			cred, err = graph.CreateRegistryCredentialFromString(credString)
-			if err != nil {
-				return err
-			}
-			credentials = append(credentials, cred)
+		task, errUnmarshal := graph.UnmarshalTaskFromString(ctx, rendered, &graph.TaskOptions{
+			DefaultWorkingDir: defaultWorkingDirectory,
+			Network:           defaultNetwork,
+			Envs:              defaultEnvs,
+			Credentials:       credentials,
+			TaskName:          taskName,
+		})
+		if errUnmarshal != nil {
+			return errors.Wrapf(errUnmarshal, "failed to unmarshal task from the string %s", rendered)
 		}
 
-		taskFinal, errUnmarshal := graph.UnmarshalTaskFromString(ctx, rendered, defaultWorkingDirectory, defaultNetwork, defaultEnvs, credentials, taskName, false)
-		if errUnmarshal != nil {
-			return errUnmarshal
+		if shouldIncludeAlias {
+			graph.ExpandCommandAliases(alias, task)
```

Comment:
>ExpandCommandAliases [](start = 9, length = 20)

the method is also called in task.NewTaskFromBytes

Reply:
yeah, not needed over there. Since now we have separated out the logic for 
SearchReplace
Expand.